src/charm.py

```diff
@@ -17,58 +19,50 @@
 )
 
 
+class CheckFailed(Exception):
+    """ Raise this exception if one of the checks in main fails. """
+
+    def __init__(self, msg, status_type=None):
+        super().__init__()
+
+        self.msg = msg
+        self.status_type = status_type
+        self.status = status_type(msg)
+
+
```

Comment:
Another tiny thing that I should have thought of earlier... Why don't we move this block to the end of the file near other helpers? It isn't the star of the show. The important thing is the operator so that should be most visible

Reply:
I really don't mind it, as the class is pretty tiny and these classes feel natural to me to be defined at the beginning. Given that this will end up a in a library it doesn't really matter.